jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+        except psutil.TimeoutExpired:
+            j.logger.debug("the start command still running after 2 sec")  # still running
+        # wait extra delay to allow any subprocess spawned from the process we just started to finish
+        # this may not needed in many cases
+        time.sleep(delay)
+        # TODO check based on command
+        if check_running(filterstr, min=n_instances):
+            j.logger.debug(f"found at least {n_instances} instances using the filter string: {filterstr}")
+            return True  # XXX should remove? None
+        else:
+            j.logger.debug(f"the required number of instances using the filter string: {filterstr} not found yet!")
+            continue
+    j.logger.error(f"could not start the required number of instances ({n_instances}) after {i} attempts.")
+    raise j.exceptions.Runtime("could not start the required number of instances.")
 
-def check_stop(cmd, filterstr, retry=1, nrinstances=0):
-    """Executes a stop command and check if it is already stopped based on filterstr
 
-    Arguments:
-        cmd {str} -- command to be executed
-        filterstr {str} -- filter string
+def check_stop(cmd, filterstr, retry=1, n_instances=0, timeout=2, delay=0.5):
+    """Executes a stop command (possibly multiple times) and check if it is already stopped based on filterstr
 
-    Keyword Arguments:
-        retry (int) -- number of retries (default: {1})
-        nrinstances (int) -- number of instances after stop (default: {0})
+    Args:
+        cmd (str): Command to be executed.
+        filterstr (str): Filter string.
+        retry (int, optional): Number of retries. Defaults to 1.
+        n_inst (int, optional): Number of instances after stop. Defaults to 0.
+        timout (int, optional): how long the function should wait for process to finish (first or parnet process started with the command)
+        delay (int, optional): how long the function should delay the checking process after the process finished or timeout exceeded (in case the first process started another one).
 
     Raises:
-        j.exceptions.RuntimeError: if nr of instances not matched
+        j.exceptions.Runtime: if number of instances not matched
     """
 
-    found = get_filtered_pids(filterstr)
     for i in range(retry):
-        if len(found) == nrinstances:
-            return
-        # print "START:%s"%cmd
-        execute(cmd, die=False)
-        time.sleep(1)
-        found = get_filtered_pids(filterstr)
-        for item in found:
-            kill(int(item), 9)
-        found = get_filtered_pids(filterstr)
-
-    if len(found) != 0:
-        raise j.exceptions.RuntimeError("could not stop %s, found %s nr of instances." % (cmd, len(found)))
-
+        if isinstance(cmd, str):
+            args = shlex.split(cmd)
+        proc = psutil.Popen(args, close_fds=True, stdout=subprocess.PIPE, stdin=subprocess.PIPE, stderr=subprocess.PIPE)
+        j.logger.debug(f"attempt no: {i}")
```

Comment:
IMHO, these inner debug logs like this can be removed after making sure everything works.